File: src/Thruster.py
#Thruster Library
#Created on 05-12-2014 by Aayush Shaurya & Akshat Agarwal

import Adafruit_BBIO.GPIO as GPIO
import Adafruit_BBIO.PWM as PWM
import time as time

class Thruster:

	global GND
	GND = "P9_1"
	COUNT_CONST=3

#	def __init__(self):
#		self.attachThruster(GND,GND,GND);

	def __init__(self, dir1_pin, dir2_pin, pwm_pin):
		self.attachThruster(dir1_pin, dir2_pin, pwm_pin);

	def attachThruster(self, dir1_pin, dir2_pin, pwm_pin):
		if(dir1_pin != GND and dir2_pin != GND and pwm_pin != GND):
			GPIO.setup(dir1_pin, GPIO.OUT)
			GPIO.setup(dir2_pin, GPIO.OUT)
			PWM.start(pwm_pin, 0)
		self.pwm = 0;
		self.speed = 0;
		self.meanSpeed = 100;
		self.dir1 = 0;
		self.dir2 = 0;
		self.dir1_pin = dir1_pin
		self.dir2_pin = dir2_pin
		self.pwm_pin = pwm_pin
		self.lockThruster();
		damping = 10;

	def getSpeed(self):
		return(pwm*100/mean_speed);

	def changePWM(self):
		if self.pwm>100:
			self.pwm=100
		elif self.pwm<-100:
			self.pwm=-100
		PWM.set_duty_cycle(self.pwm_pin, float(abs(self.pwm)))	

	def moveThruster(self, temp_dir1, temp_dir2, temp_pwm):
		self.dir1 = temp_dir1;
		self.dir2 = temp_dir2;
		self.pwm = temp_pwm;
		self.moveThruster();

	def moveThruster(self):

		if self.dir1 == 1 and self.dir2 == 0:
			self.pwm = pwm;
		elif self.dir1 == 0 and self.dir2 == 1:
			self.pwm = -pwm;

		if self.pwm >= 0:
			GPIO.output(self.dir1_pin, GPIO.HIGH)
			GPIO.output(self.dir2_pin, GPIO.LOW)
		else:
			GPIO.output(self.dir1_pin, GPIO.LOW)
			GPIO.output(self.dir2_pin, GPIO.HIGH)

		self.changePWM();

	def lockThruster(self):
			GPIO.setup(self.dir1_pin, GPIO.OUT)
			GPIO.setup(self.dir2_pin, GPIO.OUT)
			PWM.start("P8_13", 0.1)
			GPIO.output(self.dir1_pin, GPIO.HIGH)
			GPIO.output(self.dir2_pin, GPIO.HIGH)
			PWM.set_duty_cycle("P8_13", float(100))
			self.pwm = 0

	def freeThruster(self):
			GPIO.output(self.dir1_pin, GPIO.LOW)
			GPIO.output(self.dir2_pin, GPIO.LOW)
			PWM.set_duty_cycle(self.pwm_pin, float(0))
			self.pwm=0

	def setThrusterSpeed(self, temp_dir1, temp_dir2, temp_speed):
		
		if (temp_speed > 100):
			temp_speed = 100;
		if (temp_speed < -100):
			temp_speed = -100;
		
		self.speed = temp_speed;
		self.pwm = temp_speed*self.meanSpeed/100;
		self.dir1 = temp_dir1;
		self.dir2 = temp_dir2;
		self.moveThruster();

	def setThrusterSpeed(self, temp_speed):
		
		if (temp_speed > 100):
			temp_speed = 100;
		if (temp_speed < -100):
			temp_speed = -100;
		
		self.speed = temp_speed;
		self.pwm = temp_speed*self.meanSpeed/100;
		self.moveThruster();

	def changeSpeed(self, temp_speed):
		if (temp_speed > 100):
			temp_speed = 100;
		if (temp_speed < -100):
			temp_speed = -100;
		self.speed = temp_speed;
		self.changePWM(temp_speed*self.meanSpeed/100);

	def getDirection(self):
		if (self.dir1==self.dir2):
			return 0
		elif (self.dir1>self.dir2):
			return 1
		elif (self.dir2>self.dir1):
			return -1

	def isFree(self):
		return (self.getDirection()==0 and self.dir1==0)

	def isLocked(self):
		return (self.getDirection()==0 and self.dir1==1)

	def getSpeed(self):
		return(self.pwm*100/self.meanSpeed);

	def getPWM(self):
		return self.pwm;

	def startSmoothly(self, temp_speed):
		if (temp_speed > 100):
			temp_speed = 100;
		if (temp_speed < -100):
			temp_speed = -100;
		self.speed = temp_speed;
		self.startSmoothly();

	def startSmoothly(self):
		if (self.speed >= 0):
			GPIO.output(dir1_pin, GPIO.HIGH);
			GPIO.output(dir2_pin, GPIO.LOW);
		else:
			GPIO.output(dir1_pin, GPIO.LOW);
			GPIO.output(dir2_pin, GPIO.HIGH);

		i = 0;
		while (i < self.speed):
			self.changeSpeed(i);
			time.sleep(COUNT_CONST*damping/1000);
			#self.speed = i;
			i = i + COUNT_CONST
		self.pwm = self.speed*mean_speed/100

	def stopSmoothly(self):
		i=self.getSpeed()
		while (i>0):
			self.changeSpeed(i);
			time.sleep(COUNT_CONSR*damping/1000);
			i=i-COUNT_CONST;
		self.pwm = self.speed*mean_speed/100
		self.lockThruster()

File: src/thruster.cpp
#include <Thruster.h>
#include <Arduino.h>
Thruster::Thruster()									//constructor
{	
	attachThruster(-1,-1,-1);
}

Thruster::Thruster(int Dir1, int Dir2, int Pwm )		// parametrised constructor
{	
	Serial.begin(9600);
	attachThruster(Dir1,Dir2,Pwm);
	//Serial.println("Attached thruster fuck you.");
}

void Thruster::attachThruster(int Dir1,int Dir2,int Pwm)
{
	pwm_pin= Pwm;
	dir1_pin= Dir1;
	dir2_pin= Dir2;
	if(dir1!=-1 && dir2!=-1 && pwm_pin!=-1)
	{	
		pinMode(pwm_pin,OUTPUT);
		pinMode(dir1_pin,OUTPUT);
		pinMode(dir2_pin,OUTPUT);
	}
	pwm		= 0;
	speed		= 0;
	mean_speed	= 255;
	lockThruster();	//Lock the Thruster initially
	damping=10;
	}


void Thruster::moveThruster(int Pwm)			//+ve for CW and -ve for CCW.
{
	
	if(Pwm>=0)
	{	
		digitalWrite(dir1_pin,HIGH);
		digitalWrite(dir2_pin,LOW);
	}
	else
	{	
		digitalWrite(dir1_pin,LOW);
		digitalWrite(dir2_pin,HIGH);
	}
	changePWM(Pwm);
//	Serial.println("PWM=");
//	Serial.print(Pwm);
}

void Thruster::moveThruster(int Dir1,int Dir2,int Pwm)
{
	if(Dir1==1 && Dir2==0)
		moveThruster(Pwm);
	if(Dir1==0 && Dir2==1)
		moveThruster(-Pwm);
}

void Thruster::stopThruster() 				//By default stop Thruster will lock Thruster
{	
	lockThruster();
}

void Thruster::lockThruster()
{
	digitalWrite(dir1_pin,HIGH);			//case of Thruster lock
	digitalWrite(dir2_pin,HIGH);
	analogWrite(pwm_pin,255);
	pwm=0;	
}

void Thruster::freeThruster()
{
	digitalWrite(dir1_pin,LOW);			//case of Thruster free
	digitalWrite(dir2_pin,LOW);
	analogWrite(pwm_pin,0);
	pwm=0;
}
		
void Thruster::setMeanSpeed(int Speed)			//Sets the meanspeed with which Thruster moves when speed=100%	
{	
	mean_speed=Speed;
}

void  Thruster::setThrusterSpeed(int Speed)				//+ve for CW and -ve for CCW. Speed in percentage of meanspeed
{	
	if(Speed>100)
		Speed=100;
	if(Speed<-100)
		Speed=-100;
	speed=Speed;
	
	moveThruster(Speed*mean_speed/100);
}

void  Thruster::setThrusterSpeed(int Dir1,int Dir2,int Speed)
{	
	if(Speed>100)
		Speed=100;
	if(Speed<-100)
		Speed=-100;
	speed=Speed;
	moveThruster(Dir1,Dir2,Speed*mean_speed/100);
}

void Thruster::changePWM(int Pwm)					//Just to change the PWM
{
	pwm = Pwm>255 ? 255 : (Pwm < -255 ? -255 :Pwm);
	analogWrite(pwm_pin,abs(pwm));
}

void Thruster::changeSpeed(int Speed)				//Just to change the speed (In percentage)
{	
	if(Speed>100)
		Speed=100;
	if(Speed<-100)
		Speed=-100;
	speed=Speed;
	changePWM(Speed*mean_speed/100);
}

int Thruster::getDirection()
{
	dir1=digitalRead(dir1_pin);
	dir2=digitalRead(dir2_pin);
	if (dir1==dir2)
		return 0;
	else 
		return((dir1>dir2)?1:-1);
}

int Thruster::isFree()
{
	return (getDirection()==0 && dir1==0);
}

int Thruster::isLocked()
{
	return (getDirection()==0 && dir1==1);
}

int Thruster::getSpeed()
{
	return(pwm*100/mean_speed);
}

int Thruster::getPWM()
{
  return pwm;
}

 void Thruster::startSmoothly(int Speed)
 {
    if(Speed>100)
		Speed=100;
	if(Speed<-100)
		Speed=-100;
	int i;
	if(Speed>=0)
	{	
		digitalWrite(dir1_pin,HIGH);
		digitalWrite(dir2_pin,LOW);
	}
	else
	{	
		digitalWrite(dir1_pin,LOW);
		digitalWrite(dir2_pin,HIGH);
	}
	for(i=0;i<=Speed;i=i+COUNT_CONST)
	{
		changeSpeed(i);
		delay(COUNT_CONST*damping);
		speed=i;
		pwm=speed*mean_speed/100;
	}
 }
 void Thruster::stopSmoothly()
 {  int i;
	speed=getSpeed();
	for(i=speed;i>=0;i=i-COUNT_CONST)
	{
		changeSpeed(i);
		delay(COUNT_CONST*damping);
		speed=i;
		pwm=speed*mean_speed/100;
	}
	lockThruster();
 }
